File: ios/IceBloxApp/Camera/BrightnessManager.swift
import UIKit

struct BrightnessManager {
    private(set) var isDimmed = false
    private(set) var savedBrightness: CGFloat?
    private var restoreWorkItem: DispatchWorkItem?

    mutating func dim() {
        guard AppConfig.dimScreenDuringScanning else { return }
        if savedBrightness == nil {
            savedBrightness = UIScreen.main.brightness
        }
        UIScreen.main.brightness = AppConfig.dimBrightnessLevel
        isDimmed = true
    }

    mutating func restore() {
        guard isDimmed, let saved = savedBrightness else { return }
        cancelPendingRestore()
        UIScreen.main.brightness = saved
        isDimmed = false
    }

    mutating func temporarilyRestore(for duration: TimeInterval = 5.0) {
        guard isDimmed, let saved = savedBrightness else { return }
        cancelPendingRestore()
        UIScreen.main.brightness = saved
        isDimmed = false

        let workItem = DispatchWorkItem { [saved] in
            guard AppConfig.dimScreenDuringScanning else { return }
            if UIScreen.main.brightness == saved {
                UIScreen.main.brightness = AppConfig.dimBrightnessLevel
            }
        }
        restoreWorkItem = workItem
        DispatchQueue.main.asyncAfter(deadline: .now() + duration, execute: workItem)

        // Re-mark as dimmed so the pending re-dim is expected
        isDimmed = true
    }

    mutating func teardown() {
        cancelPendingRestore()
        if let saved = savedBrightness {
            UIScreen.main.brightness = saved
        }
        savedBrightness = nil
        isDimmed = false
    }

    private mutating func cancelPendingRestore() {
        restoreWorkItem?.cancel()
        restoreWorkItem = nil
    }
}

File: android/app/src/main/java/com/iceblox/app/camera/BrightnessManager.kt
package com.iceblox.app.camera

import android.app.Activity
import android.view.WindowManager
import com.iceblox.app.config.AppConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class BrightnessManager {
    var isDimmed: Boolean = false
        private set
    var savedBrightness: Float? = null
        private set
    private var restoreJob: Job? = null

    fun dim(activity: Activity?) {
        val window = activity?.window ?: return
        if (!AppConfig.DIM_SCREEN_DURING_SCANNING) return

        if (savedBrightness == null) {
            val current = window.attributes.screenBrightness
            savedBrightness = if (current < 0) 0.5f else current
        }
        setBrightness(window, AppConfig.DIM_BRIGHTNESS_LEVEL)
        isDimmed = true
    }

    fun restore(activity: Activity?) {
        val window = activity?.window ?: return
        val saved = savedBrightness ?: return
        if (!isDimmed) return

        restoreJob?.cancel()
        restoreJob = null
        setBrightness(window, saved)
        isDimmed = false
    }

    fun temporarilyRestore(
        activity: Activity?,
        scope: CoroutineScope?,
        seconds: Long = 5L
    ) {
        val window = activity?.window ?: return
        val activeScope = scope ?: return
        val saved = savedBrightness ?: return
        if (!isDimmed) return

        restoreJob?.cancel()
        setBrightness(window, saved)
        isDimmed = false

        restoreJob = activeScope.launch {
            delay(seconds * 1000)
            if (AppConfig.DIM_SCREEN_DURING_SCANNING) {
                setBrightness(window, AppConfig.DIM_BRIGHTNESS_LEVEL)
                isDimmed = true
            }
        }
        // Mark as dimmed so the pending re-dim is expected
        isDimmed = true
    }

    fun teardown(activity: Activity?) {
        restoreJob?.cancel()
        restoreJob = null
        val window = activity?.window
        val saved = savedBrightness
        if (window != null && saved != null) {
            setBrightness(window, saved)
        }
        savedBrightness = null
        isDimmed = false
    }

    private fun setBrightness(window: android.view.Window, level: Float) {
        val params = window.attributes
        params.screenBrightness = level
        window.attributes = params
    }
}
